sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/DataflowPipelineJob.java

```diff
@@ -255,19 +255,26 @@ State waitToFinish(
       }
 
       if (!hasError) {
-        // Reset the backoff.
+        // We can stop if the job is done.
+        if (state.isTerminal()) {
+          return state;
+        }
+
+        // The job is not done, so we must keep polling.
         backoff.reset();
-        // If duration is set, update the new cumulative sleep time to be the remaining
-        // part of the total input sleep duration.
+
+        // If a total duration for all backoff has been set, update the new cumulative sleep time to
+        // be the remaining total backoff duration, stopping if we have already exceeded the
+        // allotted time.
         if (duration.isLongerThan(Duration.ZERO)) {
           long nanosConsumed = nanoClock.nanoTime() - startNanos;
           Duration consumed = Duration.millis((nanosConsumed + 999999) / 1000000);
-          backoff =
-              MESSAGES_BACKOFF_FACTORY.withMaxCumulativeBackoff(duration.minus(consumed)).backoff();
-        }
-        // Check if the job is done.
-        if (state.isTerminal()) {
-          return state;
+          Duration remaining = duration.minus(consumed);
+          if (remaining.isLongerThan(Duration.ZERO)) {
+            backoff = MESSAGES_BACKOFF_FACTORY.withMaxCumulativeBackoff(remaining).backoff();
```

Comment:
`withMaxCumulativeBackoff(...)` and `withMaxBackoff(...)` should have a `Preconditions.checkArgument` that throws an appropriate error when misspecified.